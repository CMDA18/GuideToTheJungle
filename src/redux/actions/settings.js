// @flow

// Flow Types
export type TestAction = {
  type: string,
  value: string
}
export type PreloadAction = {
  type: string,
  value: string
}

// Action type constants
export const testActionType: string = 'TEST'
export const preloadActionType: string = 'PRELOAD'

// Action creators
export const createTestAction = (value: string): TestAction => ({
  type: testActionType,
  value
})

export const createPreloadAction = (value: string): PreloadAction => ({
  type: preloadActionType,
  value
})
