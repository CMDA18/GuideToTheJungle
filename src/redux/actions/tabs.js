// @flow

// Flow Types
export type TabState = {
  type: string,
  activeTab: number
}

// Action type constants
export const toggleTabActionType: string = 'TOGGLE_TAB'

// Action creators
export const createToggleTabAction = (activeTab: number): TabState => ({
  type: toggleTabActionType,
  activeTab
})
