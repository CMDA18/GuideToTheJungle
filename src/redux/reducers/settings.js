// @flow
import { preloadActionType, testActionType, type TestAction, type PreloadAction } from '../actions/settings'

// Flow Types
export type State = {
  test?: string,
  newTest?: string,
  preload?: string
}
export type Action = TestAction | PreloadAction

// Initial Reducer State when none provided
const initialState: State = { test: 'hoi' }

export default (state: State = initialState, action: Action): State => {
  switch (action.type) {
    case preloadActionType: {
      return {
        ...state,
        preload: `pre:${action.value}`
      }
    }
    case `${preloadActionType}_SUCCESS`: {
      return {
        ...state,
        preload: `pre:done:${action.value}`
      }
    }
    case testActionType: {
      return {
        ...state,
        newTest: `sync:${action.value}`
      }
    }
    case `${testActionType}_SUCCESS`: {
      return {
        ...state,
        newTest: action.value
      }
    }
    default: {
      return state
    }
  }
}
