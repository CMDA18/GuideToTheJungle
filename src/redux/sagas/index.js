import { call, put, takeLatest } from 'redux-saga/effects'

import { testActionType, preloadActionType } from '../actions/settings'

export const getAsyncContent = (target) => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(`asyncFromSaga:${target}`)
    }, 200)
  })
}

export function * asyncCallExample (action) {
  try {
    const user = yield call(getAsyncContent, action.value)
    yield put({type: `${action.type}_SUCCESS`, value: user})
  } catch (e) {
    yield put({type: `${action.type}_FAILED`, value: e.message})
  }
}

function * sagas () {
  yield takeLatest(testActionType, asyncCallExample)
  yield takeLatest(preloadActionType, asyncCallExample)
}

export default sagas
