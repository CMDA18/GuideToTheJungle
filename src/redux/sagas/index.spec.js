import { call } from 'redux-saga/effects'

import { createTestAction } from '../actions/settings'
import sagas, { asyncCallExample, getAsyncContent } from './index'

describe('async loading example', () => {
  expect.assertions(1)
  const data = getAsyncContent('tralalala')
  expect(data).resolves.toEqual('asyncFromSaga:tralalala')
})

describe('Sagas', () => {
  it('exports all sagas', () => {
    expect(sagas()).toMatchSnapshot()
  })

  it('complete an async action', () => {
    const gen = asyncCallExample(createTestAction('hoi'))

    expect(gen.next().value).toEqual(
      call(getAsyncContent, 'hoi')
    )
  })
})
