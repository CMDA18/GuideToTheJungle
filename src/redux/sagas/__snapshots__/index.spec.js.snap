// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sagas exports all sagas 1`] = `
GeneratorFunctionPrototype {
  "_invoke": [Function],
}
`;
